crates/ruff_python_ast/src/nodes.rs

```diff
@@ -999,12 +1006,155 @@ impl From<ExprFString> for Expr {
+    /// that allows modification.
+    pub(crate) fn parts_mut(&mut self) -> impl Iterator<Item = &mut FStringPart> {
+        match &mut self.inner {
+            FStringValueInner::Single(part) => Left(std::iter::once(part)),
+            FStringValueInner::Concatenated(parts) => Right(parts.iter_mut()),
+        }
+    }
+
+    /// Returns an iterator over the [`StringLiteral`] parts contained in this value.
+    ///
+    /// Note that this doesn't nest into the f-string parts. For example,
+    ///
+    /// ```python
+    /// "foo" f"bar {x}" "baz" f"qux"
+    /// ```
+    ///
+    /// Here, the string literal parts returned would be `"foo"` and `"baz"`.
+    pub fn literals(&self) -> impl Iterator<Item = &StringLiteral> {
+        self.parts().filter_map(|part| part.as_literal())
+    }
+
+    /// Returns an iterator over the [`FString`] parts contained in this value.
+    ///
+    /// Note that this doesn't nest into the f-string parts. For example,
+    ///
+    /// ```python
+    /// "foo" f"bar {x}" "baz" f"qux"
+    /// ```
+    ///
+    /// Here, the f-string parts returned would be `f"bar {x}"` and `f"qux"`.
+    pub fn f_strings(&self) -> impl Iterator<Item = &FString> {
+        self.parts().filter_map(|part| part.as_f_string())
+    }
+
+    /// Returns an iterator over all the f-string elements contained in this value.
+    ///
+    /// An f-string element is what makes up an [`FString`] i.e., it is either a
+    /// string literal or an expression. In the following example,
+    ///
+    /// ```python
+    /// "foo" f"bar {x}" "baz" f"qux"
+    /// ```
+    ///
+    /// The f-string elements returned would be string literal (`"bar "`),
+    /// expression (`x`) and string literal (`"qux"`).
+    pub fn elements(&self) -> impl Iterator<Item = &Expr> {
+        self.f_strings().flat_map(|fstring| fstring.values.iter())
+    }
+}
+
+/// An internal representation of [`FStringValue`].
+#[derive(Clone, Debug, PartialEq)]
+enum FStringValueInner {
+    /// A single f-string i.e., `f"foo"`.
+    ///
+    /// This is always going to be `FStringPart::FString` variant which is
+    /// maintained by the `FStringValue::single` constructor.
+    Single(FStringPart),
+
+    /// An implicitly concatenated f-string i.e., `"foo" f"bar {x}"`.
+    Concatenated(Vec<FStringPart>),
+}
+
+/// An f-string part which is either a string literal or an f-string.
+#[derive(Clone, Debug, PartialEq, is_macro::Is)]
+pub enum FStringPart {
+    Literal(StringLiteral),
+    FString(FString),
+}
+
+impl Ranged for FStringPart {
+    fn range(&self) -> TextRange {
+        match self {
+            FStringPart::Literal(string_literal) => string_literal.range(),
+            FStringPart::FString(f_string) => f_string.range(),
+        }
+    }
+}
+
+/// An AST node that represents a single f-string which is part of an [`ExprFString`].
+#[derive(Clone, Debug, PartialEq)]
+pub struct FString {
+    pub range: TextRange,
+    pub values: Vec<Expr>,
```

Comment:
I wonder if these should be `smallvec`... It's probably worth benchmarking in a separate PR. In practice, I'd bet the vast majority of concatenations are <= 5 elements.